resources/images/model/images.go

```diff
@@ -45,42 +50,115 @@ func NewImagesModel(
+	if metaConstructor == nil {
 		return "", controller.ErrModelMissingInputMetadata
 	}
+	if imageReader == nil {
+		return "", controller.ErrModelMissingInputArtifact
+	}
+	imageID, artifactFileID, err := i.handleArtifact(metaConstructor, imageReader)
+	// try to remove artifact file from file storage on error
+	if err != nil {
+		if cleanupErr := i.fileStorage.Delete(artifactFileID); cleanupErr != nil {
+			return "", errors.Wrap(err, cleanupErr.Error())
+		}
+	}
+	return imageID, err
+}
+
+// handleArtifact parses artifact and uploads artifact file to the file storage - in parallel,
+// and creates image structure in the system.
+// Returns image ID, artifact file ID and nil on success.
+func (i *ImagesModel) handleArtifact(
+	metaConstructor *images.SoftwareImageMetaConstructor,
+	imageReader io.Reader) (string, string, error) {
+
+	// limit just for safety
+	// max image size - 10G
+	const MaxImageSize = 1024 * 1024 * 1024 * 10
+
+	// create pipe
+	pR, pW := io.Pipe()
+	// limit reader to max image size
+	lr := io.LimitReader(imageReader, MaxImageSize)
+	tee := io.TeeReader(lr, pW)
+
+	artifactFileID := uuid.NewV4().String()
+
+	ch := make(chan error)
+	// create goroutine for artifact upload
+	//
+	// reading from the pipe (which is done in UploadArtifact method) is a blocking operation
+	// and cannot be done in the same goroutine as writing to the pipe
+	//
+	// uploading and parsing artifact in the same process will cause in a deadlock!
+	go func() {
+		err := i.fileStorage.UploadArtifact(artifactFileID, pR, ImageContentType)
+		if err != nil {
+			pR.CloseWithError(err)
+		}
+		ch <- err
+	}()
+
+	// parse artifact
+	// artifact library reads all the data from the given reader
+	metaArtifactConstructor, err := getMetaFromArchive(&tee, MaxImageSize)
+	if err != nil {
+		pW.Close()
+		if uploadResponseErr := <-ch; uploadResponseErr != nil {
+			return "", "", controller.ErrModelArtifactUploadFailed
+		}
+		return "", "", controller.ErrModelInvalidMetadata
+	}
+
+	// read the rest of the data,
+	// just in case the artifact library did not read all the data from the reader
+	_, err = io.Copy(ioutil.Discard, tee)
+	if err != nil {
+		pW.Close()
+		_ = <-ch
+		return "", "", err
+	}
 
-	if err := metaConstructor.Validate(); err != nil {
-		return "", controller.ErrModelInvalidMetadata
+	// close the pipe
+	pW.Close()
+
+	// collect output from the goroutine
+	if uploadResponseErr := <-ch; uploadResponseErr != nil {
+		return "", "", uploadResponseErr
 	}
-	if err := metaArtifactConstructor.Validate(); err != nil {
-		return "", controller.ErrModelInvalidMetadata
+
+	// validate artifact metadata
+	if err = metaArtifactConstructor.Validate(); err != nil {
+		return "", "", controller.ErrModelInvalidMetadata
 	}
+
+	// check if artifact is unique
+	// artifact is considered to be unique if there is no artifact with the same name
+	// and supporing the same platform in the system
 	isArtifactUnique, err := i.imagesStorage.IsArtifactUnique(
```

Comment:
You can read stream section by section, pull name&devicetype from headers and do early validation and send cancel to upload routine, currently you need to upload entire 2GB file just to figure out that it have been already uploaded. 

Example test from artifact lib: https://github.com/mendersoftware/mender-artifact/blob/master/reader/reader_test.go#L233